server/src/main/java/org/opensearch/action/search/SearchRequestListenerManager.java

```diff
@@ -0,0 +1,125 @@
+     *
+     * @param listeners Multiple SearchRequestOperationsListener object to add.
+     * @throws IllegalArgumentException if any input listener is null or already exists in the list.
+     */
+    public void addListeners(SearchRequestOperationsListener... listeners) {
+        for (SearchRequestOperationsListener listener : listeners) {
+            if (listener == null) {
+                throw new IllegalArgumentException("listener must not be null");
+            }
+            if (searchRequestListenersList.contains(listener)) {
+                throw new IllegalArgumentException("listener already added");
+            }
+            searchRequestListenersList.add(listener);
+        }
+    }
+
+    /**
+     * Remove a {@link SearchRequestOperationsListener} from the searchRequestListenersList,
+     *
+     * @param listener A SearchRequestOperationsListener object to remove.
+     * @throws IllegalArgumentException if the input listener is null or already exists in the list.
+     */
+    public void removeListener(SearchRequestOperationsListener listener) {
+        if (listener == null) {
+            throw new IllegalArgumentException("listener must not be null");
+        }
+        if (!searchRequestListenersList.contains(listener)) {
+            throw new IllegalArgumentException("listener does not exist in the listeners list");
+        }
+        searchRequestListenersList.remove(listener);
+    }
+
+    /**
+     * Get searchRequestListenersList,
+     *
+     * @return List of SearchRequestOperationsListener
+     * @throws IllegalArgumentException if the input listener is null or already exists in the list.
+     */
+    public List<SearchRequestOperationsListener> getListeners() {
+        return searchRequestListenersList;
+    }
+
+
+    /**
+     * Create the {@link SearchRequestOperationsListener.CompositeListener}
+     * with the all listeners enabled at cluster-level and request-level.
+     *
+     * @param searchRequest The SearchRequest object. SearchRequestListenerManager will decide which request-level listeners to add based on states/flags of the request
+     * @param logger Logger to be attached to the {@link SearchRequestOperationsListener.CompositeListener}
+     * @param perRequestListeners the per-request listeners that can be optionally added to the returned CompositeListener list.
+     * @return SearchRequestOperationsListener.CompositeListener
+     */
+    public SearchRequestOperationsListener.CompositeListener buildCompositeListener(
+        SearchRequest searchRequest,
+        Logger logger,
+        SearchRequestOperationsListener... perRequestListeners
+    ) {
+        final List<SearchRequestOperationsListener> searchListenersList = searchRequestListenersList.stream().filter(SearchRequestOperationsListener::getEnabled).collect(Collectors.toList());
+
+        Arrays.stream(perRequestListeners).forEach((listener) -> {
```

Comment:
Can this just be a plain-old `for`-loop? The `stream().forEach()` syntax isn't helping anything.

Reply:
Actually, if you apply my next comment, this loop more or less goes away -- you could iterate through the managed listeners and per-request listeners as one big stream, filtered by the enabled ones.